go/worker/upgrade/worker.go

```diff
@@ -0,0 +1,195 @@
+	upgrade "github.com/oasisprotocol/oasis-core/go/upgrade/api"
+)
+
+// Worker is a service handling governance upgrades.
+type Worker struct { // nolint: maligned
+	ctx context.Context
+
+	upgrader upgrade.Backend
+
+	consensus  consensus.Backend
+	governance governance.Backend
+	epochtime  epochtime.Backend
+
+	stopCh chan struct{}
+	quitCh chan struct{}
+
+	logger *logging.Logger
+}
+
+func (w *Worker) doWatchGovernanceUpgrades() {
+	defer close(w.quitCh)
+
+	w.logger.Info("staring governance update worker")
+
+	epochCh, sub := w.epochtime.WatchEpochs()
+	defer sub.Close()
+
+	// Wait for first block to be synced so that initial queries won't fail.
+	func() {
+		blkCh, blkSub, err := w.consensus.WatchBlocks(w.ctx)
+		if err != nil {
+			w.logger.Error("error watching consensus blocks",
+				"err", err,
+			)
+			return
+		}
+		defer blkSub.Close()
+		select {
+		case <-w.ctx.Done():
+			return
+		case <-w.stopCh:
+			return
+		case <-blkCh:
+		}
+	}()
+
+	for {
+		select {
+		case <-w.ctx.Done():
+			return
+		case <-w.stopCh:
+			return
+		case currentEpoch := <-epochCh:
+			w.logger.Debug("epoch transition event received",
+				"epoch", currentEpoch,
+			)
+			// Query governance consensus parameters.
+			params, err := w.governance.ConsensusParameters(w.ctx, consensus.HeightLatest)
+			if err != nil {
+				w.logger.Error("failed to query governance consensus parameters",
+					"err", err,
+				)
+				return
+			}
+			// Query governance pending upgrades.
+			govPendingUpgrades, err := w.governance.PendingUpgrades(w.ctx, consensus.HeightLatest)
+			if err != nil {
+				w.logger.Error("failed to query pending upgrades",
+					"err", err,
+				)
+				return
+			}
+			w.logger.Debug("governance scheduled pending upgrades",
+				"gov_pending_upgrades", govPendingUpgrades,
+				"current_epoch", currentEpoch,
+				"upgrade_cancel_min_epoch_diff", params.UpgradeCancelMinEpochDiff,
+			)
+
+			// Get pending upgrades that cannot be canceled anymore.
+			// Note: we only submit upgrades that are past the cancel epoch,
+			// to avoid having to potentially cancel upgrades.
+			var govUpgradeDescriptors []*upgrade.Descriptor
+			for _, descriptor := range govPendingUpgrades {
+				if descriptor.Epoch-currentEpoch < params.UpgradeCancelMinEpochDiff {
+					govUpgradeDescriptors = append(govUpgradeDescriptors, descriptor)
+				}
+			}
+			// Get already scheduled upgrades on the node.
+			var nodePendingUpgrades []*upgrade.PendingUpgrade
+			nodePendingUpgrades, err = w.upgrader.PendingUpgrades(w.ctx)
+			if err != nil {
+				w.logger.Error("failed to query scheduled pending upgrades",
+					"err", err,
+				)
+				return
+			}
+			w.logger.Debug("scheduled pending upgrades",
+				"gov_pending_upgrades", govUpgradeDescriptors,
+				"node_pending_upgrades", nodePendingUpgrades,
+			)
+
+			// Make sure all governance pending upgrades are scheduled.
+			for _, descriptor := range govUpgradeDescriptors {
```

Comment:
self todo: check why this is reported as not covered, although it's exercised in the governance-upgrade tests (confimed via logs)

Reply:
ah it's due to https://github.com/oasisprotocol/oasis-core/issues/3574 since nodes exit uncleanly for the upgrade (and after the upgrade is done, there's no other scheduled upgrade), should just defer the fix for that issue.